### PR TITLE
AK: Allow formatted strings in Error

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
**TL;DR: You can now do this:**
```cpp
return Error::from_string(String::formatted("Section number {} out of bounds, there are {} sections", section_number, number_of_sections));
```

---

Error can now hold either a StringView like before, or an owned String pointer. A raw pointer is necessary because of include cycles, but this will allow you to put formatted strings into Error without worrying about UAFs.

Of course, we don't implement this API for DeprecatedString to encourage the switch to new string :^)

Note that this feature is limited to userland. In the Kernel, it doesn't have much application and is difficult to OOM-harden.

A first implementation is provided for LibManual, which instantly improves error feedback of Help and man's command line arguments.

@linusg FOSDEM yakbait taken!

### CODE REVIEW

**Comment by kleinesfilmroellchen** [comment]
Drafting while I sort out the runtime failures.

File: Userland/Libraries/LibWeb/CSS/StyleValue.h [hunk 1]
```diff
@@ -150,6 +150,7 @@ namespace Filter {
 struct Blur {
     Optional<Length> radius {};
     float resolved_radius(Layout::Node const&) const;
+    bool operator==(Filter::Blur const& b) const;
```
Comment (MacDue) [inline Userland/Libraries/LibWeb/CSS/StyleValue.h:153]:
```suggestion
    bool operator==(Filter::Blur const&) const = default;
```
Now that there's an operator== on `AK::Variant` you should be able to default all these and remove the implementations from the .cpp.

File: AK/Error.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    swap(this->m_syscall, other.m_syscall);
+    swap(this->m_code, other.m_code);
+#if !defined(KERNEL)
+    if (other.m_string_literal.has<StringView>())
+        this->m_string_literal = move(other.m_string_literal.get<StringView>());
+    else if (other.m_string_literal.get<String const*>() != nullptr) {
+        this->m_string_literal = move(other.m_string_literal.get<String const*>());
+        // Make sure the other Error doesn't destruct the string!
+        other.m_string_literal.get<String const*>() = nullptr;
+    }
```
Comment (trflynn89) [inline AK/Error.cpp:23]: If one branch in an if/else-if chain requires braces, all chains should have braces.

File: AK/Error.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+#else
+    swap(this->m_string_literal, other.m_string_literal);
+#endif
+}
+
+Error::Error(Error const& other)
+    : m_code(other.m_code)
+    , m_syscall(other.m_syscall)
+{
+
```
Comment (trflynn89) [inline AK/Error.cpp:33]: Remove empty newline

File: AK/Error.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+#include "Error.h"
+#include "String.h"
+
+namespace AK {
+
+Error::Error(Error&& other)
+{
+    swap(this->m_syscall, other.m_syscall);
+    swap(this->m_code, other.m_code);
+#if !defined(KERNEL)
```
Comment (trflynn89) [inline AK/Error.cpp:16]: IMO ifdef soup is a little less soupy with the "true" case being first:
```c++
#if defined(KERNEL)
// kernel stuff
#else
// userland
#endif
```

File: AK/Error.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    if (other.m_string_literal.has<StringView>())
+        this->m_string_literal = move(other.m_string_literal.get<StringView>());
+    else if (other.m_string_literal.get<String const*>() != nullptr) {
+        this->m_string_literal = move(other.m_string_literal.get<String const*>());
```
Comment (trflynn89) [inline AK/Error.cpp:20]: Both of these `move`s don't do anything and I'd bet clangd even says to remove them

File: AK/Error.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    : m_code(other.m_code)
+    , m_syscall(other.m_syscall)
+{
+
+#if !defined(KERNEL)
+    if (other.m_string_literal.has<StringView>())
+        this->m_string_literal = other.m_string_literal.get<StringView>();
+    else if (other.m_string_literal.get<String const*>() != nullptr) {
+        this->m_string_literal = new (nothrow) String { *other.m_string_literal.get<String const*>() };
+    }
```
Comment (trflynn89) [inline AK/Error.cpp:39]: Same here re: braces in an if chain, but these just aren't needed

File: AK/Error.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    if (m_string_literal.has<String const*>() && m_string_literal.get<String const*>() != nullptr)
+        delete m_string_literal.get<String const*>();
+#endif
+}
+
+#if !defined(KERNEL)
+Error::Error(String const* owned_string_literal)
+    : m_string_literal(owned_string_literal)
+{
+    VERIFY(m_string_literal.has<String const*>());
```
Comment (trflynn89) [inline AK/Error.cpp:62]: This is a bit over-the-top defensive IMO. It reads like:
```c++
bool b = true;
VERIFY(b == true);
```

File: AK/Error.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+Error Error::from_string(ErrorOr<String>&& string_or_error, Optional<int> code)
+{
+    // We intentionally do not return the String creation error here, since that is probably unrelated to the actual error.
+    if (string_or_error.is_error()) {
+        if (code.has_value())
+            return { code.value() };
+        return { ""sv };
```
Comment (trflynn89) [inline AK/Error.cpp:78]: Not sure this is a good fallback, we get no information here and will have no idea what the error is / where it came from. Not sure what to do though, short of adding a `SourceLocation` to `Error`.

File: AK/Error.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Error.h"
+#include "String.h"
```
Comment (trflynn89) [inline AK/Error.cpp:8]: <AK/Error.h> and <AK/String.h>.

File: AK/Error.cpp (hunk 5 above)
Comment (kleinesfilmroellchen) [reply AK/Error.cpp:20]: clangd actually doesn't say so, I was just trying to be careful about semantics here since we can easily cause memory bugs if we don't do this right.

Comment (trflynn89) [reply AK/Error.cpp:20]: Maybe whatever enables this warning depends on your editor's LSP plugin, but I just checked your branch out and indeed get:

Screenshot from 2023-02-08 14-54-51

File: AK/Error.cpp (hunk 8 above)
Comment (kleinesfilmroellchen) [reply AK/Error.cpp:78]: There's no general-purpose errno, so we have no good option here. I don't want to propagate the string error since that's most likely unrelated.

File: AK/Error.cpp [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+{
+    // We intentionally do not return the String creation error here, since that is probably unrelated to the actual error.
+    if (string_or_error.is_error()) {
+        if (code.has_value())
+            return { code.value() };
+        return { ""sv };
+    }
+
+    // Note that we can safely put the nullptr into the new Error, since string_literal handles that case.
+    auto const* heap_string = new (nothrow) String { string_or_error.release_value() };
```
Comment (alimpfard) [inline AK/Error.cpp:79]: :mortified_face: why are we allocating a string, and worse, why are we throwing the error into the ocean if this allocation fails?

"I couldn't allocate a new one, so a partially filled error is all you get"...?

File: AK/Error.h [hunk 11]
```diff
@@ -19,12 +21,27 @@
...
+template<typename T, typename E>
+class [[nodiscard]] ErrorOr;
+
 class Error {
 public:
+    Error(Error&&);
+    Error(Error&);
+    Error(Error const&);
+
+    ~Error();
```
Comment (alimpfard) [inline AK/Error.h:33]: Have you checked the performance impact of this? Error not being a POD type will cause all sorts of optimisations to go away.

File: AK/Error.cpp [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    : m_code(other.m_code)
+    , m_syscall(other.m_syscall)
+{
+#if defined(KERNEL)
+    this->m_string_literal = other.m_string_literal;
+#else
+    if (other.m_string_literal.has<StringView>())
+        this->m_string_literal = other.m_string_literal.get<StringView>();
+    else if (other.m_string_literal.get<String const*>() != nullptr)
+        this->m_string_literal = new (nothrow) String { *other.m_string_literal.get<String const*>() };
```
Comment (alimpfard) [inline AK/Error.cpp:39]: "and this empty error belongs to...someone"

Comment (kleinesfilmroellchen) [reply AK/Error.cpp:39]: Why empty? We only create empty errors via static StringViews.

File: AK/Error.cpp (hunk 8 above)
Comment (kleinesfilmroellchen) [reply AK/Error.cpp:75]: I will enforce a StringView fallback.

**Comment by kleinesfilmroellchen** [comment]
This problem sucks. You can take this branch and do whatever, but I won't help you.
